Operator + Forwarder docs (#1101)
smartcontractkit/documentation: src/pages/chainlink-nodes/contracts/forwarder.md

To solve these challenges, we introduced two major features that will allow node operators to set up different transaction-sending strategies more securely while lowering their infrastructure costs:

- Chainlink nodes support multiple EOAs.
- [Forwarder](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.7/AuthorizedForwarder.sol) contracts. Forwarder contracts allow a node operator to manage multiple EOAs and make them look like a single address. If you use a web2 analogy, Forwarder contracts act like a reverse proxy server, where the user is served by the same address (reverse proxy) and does not see where(server) the traffic is coming from.

Combining multiple EOAs setups and Forwarder contracts opens up a lot of flexibility and security in terms of design:

- Node operators can expand horizontally using multiple EOAs. They can deploy one or multiple Forwarder contracts for these EOAs. These combinations of EOAs + forwarders offer a lot of flexibility for setting up different transaction-sending strategies.
- Node operators can support different job types (OCR, VRF, API request..Etc) on the same node, reducing maintenance and infrastructure costs.

> - Security-wise, Forwarder contracts distinguish between Owner and Authorized Senders. Authorized senders are hot wallets (Chainlink nodes' EOAs). The owner is expected to be a more secure address, such as a hardware wallet or a multisig wallet. The owner is responsible for changing the Authorized senders' list (e.g., if a node is compromised).

Reviewer: I found this. Our expectation is that forwarder is going to be owned by an operator, the operator is the one that's going to be owned by a HW or Multisig

---
layout: ../../../layouts/MainLayout.astro
section: nodeOperator
date: Last Modified
title: "Forwarder"
---

In the EVM world, [Externally-owned accounts'](https://ethereum.org/en/developers/docs/accounts/) transactions are confirmed sequentially. Chainlink nodes that use a single Externally-owned account (EOA) per chain face several important challenges:

- Transactions are broadcast through a single FIFO queue. Hence the throughput is limited to a single lane. A low throughput is terrible for the overall User experience as concurrent clients have to wait to get their requests fulfilled (the more clients, the more they have to wait).
- Transactions are not executed by priority. For example, let's consider there are two transactions in a row: the first one is a fulfillment of an API request(e.g., get the winner of the Fifa world cup 2022), and the second one is an _ETH/USD_ price feed update used by multiple DeFi protocols. Relying on a single EOA makes the 2nd transaction (price feed update) be confirmed after the 1st one (get the Fifa world cup winner).

> - A stuck transaction (e.g., because the gas price was too low) will cause all the subsequent transactions to remain pending.

Author: If the gas limit is too low, then the transaction fails. Here, I'm taking an example where the gas price (paid to miners) is not high enough, so they are not incentivized to include the transaction in the next blocks. This is just an example, that's why I put it in brackets (transaction can remain pending for other reasons)

> - A stuck transaction (e.g., because the gas price was too low) will cause all the subsequent transactions to remain pending.

Reviewer 2: I understand now. I edited my suggestion. It's more verbose, but perhaps it is more clear? Generally I avoid parenthesis or e.g. types of asides to keep sentences more concise and clear even if it requires an additional clarifying sentence.

- As a workaround, some node operators deploy multiple Chainlink nodes per chain. While this allows them to set up different strategies (e.g., one Chainlink node for price feed and another to fulfill API requests), this comes with an overhead in infrastructure and maintenance costs.

To solve these challenges, we introduced two major features that will allow node operators to set up different transaction-sending strategies more securely while lowering their infrastructure costs:

- Chainlink nodes support multiple EOAs.
- [Forwarder](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.7/AuthorizedForwarder.sol) contracts. Forwarder contracts allow a node operator to manage multiple EOAs and make them look like a single address. If you use a web2 analogy, Forwarder contracts act like a reverse proxy server, where the user is served by the same address (reverse proxy) and does not see where(server) the traffic is coming from. To do so, nodes call the [forward](#forward) function on the Forwarder contract.

Combining multiple EOAs setups and Forwarder contracts opens up a lot of flexibility and security in terms of design:

> - Node operators can expand horizontally using multiple EOAs. They can deploy one or multiple Forwarder contracts for these EOAs. These combinations of EOAs + forwarders offer a lot of flexibility for setting up different transaction-sending strategies.

Reviewer 2: ```suggestion
- Node operators can expand horizontally using multiple EOAs. They can deploy one or multiple Forwarder contracts for these EOAs. The combination of EOAs and forwarders offers a lot of flexibility for setting up different pipelines for handling transactions.
```

Is it more accurate to say pipelines or pathways instead of strategies? Strategy is more about planning than handling transactions.

Author: Thanks, @dwightjl; that makes more sense indeed: Strategy is about planning, and Pipeline is about execution.

> - Node operators can support different job types (OCR, VRF, API request..Etc) on the same node, reducing maintenance and infrastructure costs.

Reviewer 2: ```suggestion
- Node operators can support different job types (OCR, VRF, API request..Etc) on the same node, which reduces maintenance and infrastructure costs.
```

> - Security-wise, Forwarder contracts distinguish between Owner and Authorized Senders. Authorized senders are hot wallets (Chainlink nodes' EOAs). The owner is responsible for changing the Authorized senders' list (e.g., if a node is compromised).

Reviewer 2: ```suggestion
- Security-wise, forwarder contracts distinguish between owner accounts and authorized sender accounts. Authorized senders are hot wallets such as the EOAs of a Chainlink node. If a node is compromised, the owner is responsible for changing the authorized senders list.
```

> - Node operators do not need to manually compile and deploy [Operator](/chainlink-nodes/contracts/operator) and [Forwarder](/chainlink-nodes/contracts/forwarder) contracts. They can deploy them directly from the [OperatorFactory](/chainlink-nodes/contracts/operatorfactory) by calling the [deploynewoperatorandforwarder](/chainlink-nodes/contracts/operatorfactory#deploynewoperatorandforwarder) function. From a design perspective, the owner of a Forwarder contract is an [Operator](/chainlink-nodes/contracts/operator) contract. The owner of the Operator contract is a more secure address, such as a hardware wallet or a multisig wallet. Therefore, node operators can manage a set of Forwarder contracts through an Operator contract using a secure account such as hardware or a multisig wallet.

Reviewer 2: ```suggestion
- Node operators do not need to manually compile and deploy [operator](/chainlink-nodes/contracts/operator) and [forwarder](/chainlink-nodes/contracts/forwarder) contracts. They can deploy them directly from the [operator factory](/chainlink-nodes/contracts/operatorfactory) by calling the [deploynewoperatorandforwarder](/chainlink-nodes/contracts/operatorfactory#deploynewoperatorandforwarder) function. From a design perspective, the owner of a forwarder contract is an [operator](/chainlink-nodes/contracts/operator) contract. The owner of the operator contract is usually a more secure address with keys stored in a hardware wallet or protected by a multisig. Node operators can manage a set of forwarder contracts through an operator contract.
```

If you deploy from the operator factory, is the address that requests the deployment the owner? Is this more of a recommendation to use a hardware wallet or multisig when requesting deployments from the factory?

Author: Yes @dwightjl: the owner deploys the contracts using the factory. Therefore, we recommend that the owner's private key is protected by strong security mechanisms (e.g., hardware wallet). Alternatively, the owner could be a multisig where each signer protects their private key using a hardware wallet.

## Api Reference

> The Forwarder contract inherits [AuthorizedReceiver](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.7/AuthorizedReceiver.sol) and [ConfirmedOwnerWithProposal](https://github.com/smartcontractkit/chainlink/edit/develop/contracts/src/v0.7/ConfirmedOwnerWithProposal.sol). Read [AuthorizedReceiver](/chainlink-nodes/contracts/receiver) and [ConfirmedOwnerWithProposal](/chainlink-nodes/contracts/ownership) API references.

Reviewer 2: ```suggestion
The forwarder contract inherits [AuthorizedReceiver.sol](https://github.com/smartcontractkit/chainlink/blob/develop/contracts/src/v0.7/AuthorizedReceiver.sol) and [ConfirmedOwnerWithProposal.sol](https://github.com/smartcontractkit/chainlink/edit/develop/contracts/src/v0.7/ConfirmedOwnerWithProposal.sol). Read the [Receiver](/chainlink-nodes/contracts/receiver) and [Ownership](/chainlink-nodes/contracts/ownership) API references to learn more.
```
